File: src/main/scala/org/anz/codechallenge/validators/ColumnValidator.scala
package org.anz.codechallenge.validators

import com.google.gson.Gson
import org.anz.codechallenge.schema.{Metadata, Schema, Tag}
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.functions.col

import scala.io.Source

/**
  * Column validator
  * @param inputMetadata - Input metadata representing metadata info of input file
  * @param file_schema - File schema representing schema of the input file
  * @param tagFile - Tagfile representing he file name and record count
  * @param dataFrame - Input file read as dataframe which needs validation
  */
class ColumnValidator (
                        inputMetadata: Metadata,
                        file_schema: Schema,
                        tagFile: Tag,
                        dataFrame: DataFrame
                      ) extends Validator {

  /**
    * Perform column validation
    * @return - status of validation
    */
  override def validate(): String = {
    println("Performing column validation")

    val expectedColNames = file_schema.columns.map(fs => fs.name)
    val actualColumns = dataFrame.columns

    println("Actual column names "+actualColumns.mkString)
    println("Expected column names "+expectedColNames.mkString)

    val result = expectedColNames.sameElements(actualColumns)

    val status = if(result)  "0" else "4"

    println("Column validation status is "+status)
    status
  }
}

File: src/main/java/org/anz/codechallenge/validators/ColumnValidator.java
package org.anz.codechallenge.validators;

import org.anz.codechallenge.filedetails.ContentParams;
import org.anz.codechallenge.filedetails.FileContent;
import org.anz.codechallenge.schema.JSONSchema;
import org.anz.codechallenge.schema.Schema;
import org.anz.codechallenge.tags.DelimitedTag;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;

import java.util.Arrays;

public class ColumnValidator implements Validator {

    private FileContent fileContent;

    public ColumnValidator(FileContent fileContent) {
        this.fileContent = fileContent;
    }

    @Override
    public String validate() {
        System.out.println("Performing column validation");

        Schema fileSchema = fileContent.getFileMetadata().getFileSchema();
        boolean result = false;

        if(fileSchema instanceof JSONSchema) {
            JSONSchema jsonSchema = (JSONSchema) fileSchema;
            String[] expectedColNames = jsonSchema.getColumns().stream().map(fs -> fs.getName()).toArray(String[]::new);
            String[] actualColumns = fileContent.getDataframe().columns();

            System.out.println("Actual column names " + actualColumns);
            System.out.println("Expected column names " + expectedColNames);

            result = Arrays.equals(expectedColNames, actualColumns);
        }
        String status = (result) ? "0" : "4";

        System.out.println("Column validation status is "+status);

        return status;
    }
}
